### ISSUE TITLE
Upgrade to latest version of CSS (including Display builder)

### ISSUE THREAD

**Issue by John-Holt-Tessella** [ISISComputingGroup/IBEX#3087]
Upgrade E4 client to use the latest version of CSS (4.5) and databrowser3. As part of this opis should be convereted to Display builder unless there is a compelling reason not to.

Should be done before #2875.

**Comment by AdrianPotter** [comment]
I have investigated updating to display builder and am recommending that we don't do it at present time.

# Deployment state

At the moment, display builder is a separate repo of Kay Kasemir and not integrated as part of CSStudio, as of release 4.5.7. This includes both display builder and databrowser3:

https://github.com/kasemir/org.csstudio.display.builder
https://github.com/kasemir/org.csstudio.display.builder/tree/master/databrowser3

# CSS IDE integration

In order to edit display builder (*.bob) files, we'd need a version of CSS that has it integrated. I initially tried using the latest packaged IDE (4.5.7) available from the main CSS website. That doesn't have display builder integrated, and my attempt to install it failed with a variety of `javaFX` issues.

I subsequently used the SNS version of CSS, the latest version being 4.5.0. This does have display builder integrated and could be used to develop *.bob files, or work with converted *.opi files:

https://ics-web.sns.ornl.gov/css/

# E4 integration

I did manage to successfully integrate display builder into our E4 deployment. I am not proposing merging the code but I will leave it on the branch:

https://github.com/ISISComputingGroup/ibex_gui/tree/Ticket3087_Upgrade_to_display_builder

It may be useful in the future. It took a little bit of time to figure out the entry point for display builder. I introduced a new class, `DisplayBuilderView` that extended from `RuntimeViewPart`. This could then largely be used in place of the existing `OPIView` with some minor modification to macros and paths.

The harder part of the integration was setting up the required runtime environment. The following changes were necessary, and are all contained on the branch above except for the VM arguments:

- In Eclipse under Project/Run Configurations/Arguments/VM arguments add `Dorg.osgi.framework.bundle.parent=ext -Dosgi.framework.extensions=org.eclipse.fx.osgi` as documented on the display builder repo
- Add the necessary javaFX and display builder dependencies to the target platform

![image](https://user-images.githubusercontent.com/10273443/38869763-a6f5b43a-4243-11e8-82ec-c450ad908f52.png)

- Add the necessary dependencies to the `MANIFEST.mf` for the modified plugins. Note that unless you re-export dependencies then you'll need to add the dependencies to `*.ui.devicescreens` etc. The snapshot below shows the current working dependencies on my branch for `*.ui.devicescreens`. Again, the full list of changes can be accessed from the branch above:

![image](https://user-images.githubusercontent.com/10273443/38869844-ef69a136-4243-11e8-8173-057fb82d7387.png)

With those changes, it should be possible to open opis in E4 as display builder screens.

I had a number of issues with exceptions telling me about missing class definitions regarding `javaFX`. If you get those, be sure that the VMargs detailed above have been successfully added to your run configuration, and the target platform has been updated accordingly

# OPI to BOB conversion

When opening OPI files with display builder, it will automatically try and convert them to display builder views. This is reasonably good, but produces screens that do not look as good as the originals. There are a few notable bits of functionality I've seen so far that we would need to address:

- Scripts no longer work. Display builder doesn't let you use the existing `PVUtil` or `ConsoleUtil`. You need to use the new one. That is we'd have to change `from org.csstudio.opibuilder.scriptUtil import PVUtil` to `from org.csstudio.display.builder.runtime.script import PVUtil`. Notably there is no longer a `ConsoleUtil` in that package, but you can instead use certain methods like logging or message boxes from `ScriptUtil`. The interface has also been changed significantly for `PVUtil` (e.g. no `isConnected()) so our existing scripts would need updating. I don't see an efficient way of automating it. Notably the source is here: `https://github.com/kasemir/org.csstudio.display.builder/tree/master/org.csstudio.display.builder.runtime/src/org/csstudio/display/builder/runtime/script` which is useful for figuring out how to access the classes.
- Non-native widgets are converted successfully but our usage means this isn't always correct. For instance, the CryoValve OPI uses a text update widget in place of labels. When it is converted, it loses its internal values and appears as a text input box.

Based on these, I think when we migrate to display builder we would need to reformat all of our existing OPIs. Not only is there a noticeable drop in aesthetic quality from the automatic conversion, most OPIs have at least some features that need tweaking owing to the conversion:

- Confirm they all work [must]
- Modify scripts and incorrectly converted widgets [must]
- Eliminate the copious number of reported conversion errors on opening OPIs [nice to have]
- Reverse aesthetic degradation [nice to have]

As an example, this is what the automatically converted Eurotherm looks like:

![image](https://user-images.githubusercontent.com/10273443/38870406-9dabcdb8-4245-11e8-812b-6d6a893521bb.png)

Notably the error fields appear as hashes because the underlying scripts fail. The toggle buttons don't work for the same reason.

# Summary

- Display builder hasn't yet been integrated into core CSS
- Including it before then involves a significant overhead (I haven't even looked at whether it works from the build script yet)
- When and if we do migrate, it will require significant development effort

# Conclusion

We shouldn't take on the extra work and complexity of this migration at least until the features are part of the official CSS (rather than SNS-CSS) release.

For the time being I will create a PR for updating to the latest version of CSS, since it's straightforward and worth staying up-to-date

**Comment by kjwoodsISIS** [comment]
We need to consider what we will deliver in terms of an E4 based client this summer.  If integration of Display Builder is significant extra work, then I am not in favour of doing so until we have a working, deployed E4 client.

**Comment by AdrianPotter** [comment]
Agreed. I've created a PR for doing basic upgrades of the dependencies. I'm just sorting out PR/local build failure

**Comment by kjwoodsISIS** [comment]
@AdrianPotter - above you say "Add the necessary javaFX and display builder dependencies ...".  However, [this article](https://www.theregister.co.uk/2018/03/09/java_release_train_qcon/) from TheRegister says "JavaFX – originally Sun's answer to Flash and Silverlight, this will be gone in JDK 11".  JDK 11 will be released in September 2018 (yes, you read that right!).  Is DisplayBuilder holed below the water line?

**Comment by AdrianPotter** [comment]
It's not yet clear in my opinion. Oracle says it will let JavaFX move at its own pace and allow greater ease of 3rd party contributions. The cynic in me says they're casting it adrift. In the case of the latter, then it raises serious questions about the direction display builder is heading in. I would certainly hold back on jumping on board until the long-term plan is discussed.

**Comment by kjwoodsISIS** [comment]
The description in this ticket says "opis should be convereted to Display builder unless there is a compelling reason not to".  I think we have a compelling reason not to convert to DisplayBuilder.  Does anyone disagree?

**Comment by AdrianPotter** [comment]
Yes. I'm just trying to work out why it doesn't build on @Tom-Willemsen's machine (my machine, build server are ok).
